virtcontainers/qemu.go

```diff
@@ -423,6 +423,17 @@ func (q *qemu) setupTemplate(knobs *govmmQemu.Knobs, memory *govmmQemu.Memory) g
 	return incoming
 }
 
+func (q *qemu) setupFileBackedMem(knobs *govmmQemu.Knobs, memory *govmmQemu.Memory) {
```

Comment:
There are now 3 users:
1. VM templating, see setupTemplate()
2. virtio-fs
3. Explicit file_mem_backend in configuration.toml

Additionally, all other vhost-user devices also require this setting if hugepages is disabled.  Perhaps it should be enabled automatically for all vhost-user devices, not just virtio-fs?

When VM templating is enabled together with virtio-fs the MemoryPath is ignored in favor of FileBackedMemRootDir.  Did you check that VM templating still works in this case?

Reply:
@ganeshmaharaj We also have a gap around memory hotplug. Our memory hotplug model should also switch over to file_mem_backend for all hotplugged dimms also when this is flipped. All the vhost-user backends may or may not tolerate this, but we should ensure we do not end up with a mix of file backed and no file backed DIMMs.

Reply:
@ganeshmaharaj just to be clear, when you flip this flag every DIMM should be file backed. (This also applies to huge pages). So this will be a good time to fix hotplug handling overall. It can be in a separate PR as this feature is experimental in this release. But before we move it out of experimental we need to make the change as kubernetes will break as it relies heavily on memory hotplug support today.

/cc @dagrh

Reply:
@mcastelino It is a common use case for vm templating to mix file backed and non-file backed DIMMs. When we unmix them, vm templating's initial memory path should be excluded since it is always file backed.

Reply:
@bergwolf fair enough. However vhost-user backends will break if we do not make them all file based. So when vhost-user is used, we should default all DIMMs to file based. When it is not in use we should allow the mixing.

However it may just be easier to go one way or the other. (If the overhead is marginal)

Reply:
@mcastelino Once vm template's initial memory path is excluded, I'm all for it to never mix the two.

Reply:
Have fixed the choices on multiple scenarios. VM Templating gets the high priority off all. If VM templating is not enabled, the user's filebacking location will be used and fallback is `/dev/shm`. @stefanha I was not able to get VM templating to work with virtio-fs. That is now next in the list of things to tackle once this patch lands. @mcastelino @bergwolf so will memory hotplug and removing the mixing of memory backends.

Reply:
@bergwolf @ganeshmaharaj in summary if VM templating is enabled then virtio-fs cannot be used. Actually any vhost-user implementation cannot be used.

If I understand right this is due to the fact that the file backing the templated RAM is setup with share=off. Is that a correct summary. So this is independent of hotplug right?

Can we add this as a known limitation to the documentation?

Reply:
@ganeshmaharaj @mcastelino I think it is a problem due to kata/govmm implementation. vm template is using its own dimm, and virtio-fs is using its own dimm. Their `share` property is independent to each other. I don't see why we cannot make them work together.